src/shared/mount-util.c

```diff
@@ -490,52 +490,81 @@ int mount_move_root(const char *path) {
         return RET_NERRNO(chdir("/"));
 }
 
-int mount_pivot_root(const char *path) {
+int mount_pivot_root(const char *path, MountAttrPropagationType type) {
         _cleanup_close_ int fd_oldroot = -EBADF, fd_newroot = -EBADF;
+        unsigned int flags;
+        const char *prop_name;
 
         assert(path);
 
-        /* pivot_root() isn't currently supported in the initramfs. */
-        if (in_initrd())
-                return mount_move_root(path);
-
-        fd_oldroot = open("/", O_PATH|O_DIRECTORY|O_CLOEXEC|O_NOFOLLOW);
-        if (fd_oldroot < 0)
-                return log_debug_errno(errno, "Failed to open old rootfs");
-
-        fd_newroot = open(path, O_PATH|O_DIRECTORY|O_CLOEXEC|O_NOFOLLOW);
-        if (fd_newroot < 0)
-                return log_debug_errno(errno, "Failed to open new rootfs '%s': %m", path);
-
-        /* Change into the new rootfs. */
-        if (fchdir(fd_newroot) < 0)
-                return log_debug_errno(errno, "Failed to change into new rootfs '%s': %m", path);
-
-        /* Let the kernel tuck the new root under the old one. */
-        if (pivot_root(".", ".") < 0)
-                return log_debug_errno(errno, "Failed to pivot root to new rootfs '%s': %m", path);
-
-
-        /* At this point the new root is tucked under the old root. If we want
-         * to unmount it we cannot be fchdir()ed into it. So escape back to the
-         * old root. */
-        if (fchdir(fd_oldroot) < 0)
-                return log_debug_errno(errno, "Failed to change back to old rootfs: %m");
-
-        /* Note, usually we should set mount propagation up here but we'll
-         * assume that the caller has already done that. */
+        switch (type) {
+        case PROPAGATION_INHERIT:
+                flags = 0;
+                prop_name = "inherited";
```

Comment:
We usually introduce a string table and define a simple function with `DEFINE_STRING_TABLE_LOOKUP_TO_STRING()` macro.

Reply:
Yeah, I thought about this but thought it was overkill for this case.